library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +95,79 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        // This is a helper struct to allocate the header with proper alignment:
+        // Header **end** must be aligned to the value.
+        // We insert the padding not after the `header` field, but before the `header` field.
+        // Allocation layout is be:
+        // ```
+        // [ padding | metadata | ZST value ]
+        // ```
+        #[repr(C)]
+        struct DynZstAlloc<T, Dyn: ?Sized> {
+            header: MaybeUninit<<Dyn as Pointee>::Metadata>,
+            value: MaybeUninit<T>,
+        }
+
+        impl<'a, T: 'a, Dyn: ?Sized + 'a> DynZstAlloc<T, Dyn>
+        where
+            T: Unsize<Dyn>,
+        {
+            const fn value_offset() -> usize {
+                mem::offset_of!(DynZstAlloc<T, Dyn>, value)
+            }
+
+            const ALLOC: &'a DynZstAlloc<T, Dyn> = {
+                // SAFETY: this is compile time evaluation.
```

Comment:
What's that supposed to mean? UB at compile-time is still UB! See https://rust-lang.github.io/rfcs/3016-const-ub.html.

Reply:
I thought there's no UB at compile time. I'll change the comment.

Reply:
Yeah there definitely is UB at compile-time.

If you can find out where that misconception comes from, maybe we can improve the docs to reduce the risk of other people running into it.